kv/etcd/etcd.go

```diff
@@ -0,0 +1,292 @@
+// THE SOFTWARE.
+
+package etcd
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/golang/protobuf/proto"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+	"github.com/uber-go/tally"
+	"golang.org/x/net/context"
+)
+
+var noopCancel func()
+
+// NewStore creates a kv store based on etcd
+func NewStore(etcd *clientv3.Client, opts Options) kv.Store {
+	scope := opts.InstrumentsOptions().MetricsScope()
+	return &client{
+		opts:       opts,
+		kv:         etcd.KV,
+		watcher:    etcd.Watcher,
+		watchables: map[string]kv.ValueWatchable{},
+		retrier:    xretry.NewRetrier(opts.RetryOptions()),
+		logger:     opts.InstrumentsOptions().Logger(),
+		m: clientMetrics{
+			etcdGetError:    scope.Counter("etcd-get-error"),
+			etcdPutError:    scope.Counter("etcd-put-error"),
+			etcdTnxError:    scope.Counter("etcd-tnx-error"),
+			etcdWatchCreate: scope.Counter("etcd-watch-create"),
+			etcdWatchError:  scope.Counter("etcd-watch-error"),
+			etcdWatchReset:  scope.Counter("etcd-watch-reset"),
+		},
+	}
+}
+
+type client struct {
+	sync.RWMutex
+
+	opts       Options
+	kv         clientv3.KV
+	watcher    clientv3.Watcher
+	watchables map[string]kv.ValueWatchable
+	retrier    xretry.Retrier
+	logger     xlog.Logger
+	m          clientMetrics
+}
+
+type clientMetrics struct {
+	etcdGetError    tally.Counter
+	etcdPutError    tally.Counter
+	etcdTnxError    tally.Counter
+	etcdWatchCreate tally.Counter
+	etcdWatchError  tally.Counter
+	etcdWatchReset  tally.Counter
+}
+
+func (c *client) Get(key string) (kv.Value, error) {
+	ctx, cancel := c.context()
+	defer cancel()
+
+	r, err := c.kv.Get(ctx, c.opts.KeyFn()(key))
+	if err != nil {
+		c.m.etcdGetError.Inc(1)
+		return nil, err
+	}
+
+	if r.Count == 0 {
+		return nil, kv.ErrNotFound
+	}
+
+	if r.Count > 1 {
+		return nil, fmt.Errorf("received %d values for key %s, expecting 1", r.Count, key)
+	}
+
+	return kv.NewValue(r.Kvs[0].Value, int(r.Kvs[0].Version)), nil
+}
+
+func (c *client) Watch(key string) (kv.ValueWatch, error) {
+	c.Lock()
+	watchable, ok := c.watchables[key]
+	if !ok {
+		watchChan := c.watcher.Watch(
+			context.Background(),
+			c.opts.KeyFn()(key),
+			clientv3.WithProgressNotify(),
+			// Receive initial notification once the watch channel is created
+			clientv3.WithCreatedNotify(),
+		)
+		c.m.etcdWatchCreate.Inc(1)
+
+		watchable = kv.NewValueWatchable()
+		c.watchables[key] = watchable
+
+		go func() {
+			ticker := time.Tick(c.opts.WatchChanCheckInterval())
+
+			for {
+				select {
+				case r, ok := <-watchChan:
+					if ok {
+						c.processNotification(r, watchable, key)
+					} else {
+						c.logger.Warnf("etcd watch channel closed on key %s, recreating a watch channel", key)
+
+						// avoid recreating watch channel too frequently
+						time.Sleep(c.opts.WatchChanResetInterval())
+
+						watchChan = c.watcher.Watch(
```

Comment:
Yeah that's exactly what I wanted, I'm trying be defensive for the case that Get() call followed by the watch notification failed with all retries on a watch update, without WithProgressNotify, I'll wait until next event happens on the key for this go routine to try another Get(), which could be really long. now with WithProgressNotify, I can at least try another Get() every 10 mins, which does not add much work load but make sure it has another chance to check for any updates every 10 mins.

As you can see I use watch updates as notifications only, rather than looking into what kind of update it is, be it update or create or delete. Is there any option on the Watch call to not get those information, only get notifications? If not, probably worth adding one?